File: Eunleelee/2024-01/1월 23일/메뉴_리뉴얼.py
from itertools import combinations
from collections import Counter

def solution(orders, course):
    _list = []
    for order in orders:
        for c in course:
            _list += list(combinations(sorted(order), c))

    _dict = {k: [2] for k in course}
    for k, v in Counter([''.join(v) for v in _list]).items():
        if _dict[len(k)][0] == v:
            _dict[len(k)].append(k)
        elif _dict[len(k)][0] < v:
            _dict[len(k)] = [v, k]

    answer = []
    for v in _dict.values():
        answer += v[1:]
    return sorted(answer)

File: yht0827/1월/1월 23일/메뉴 리뉴얼.java
import java.io.*;
import java.util.*;

class Solution {
	private static List<String> combination;

	public String[] solution(String[] orders, int[] course) {
		String[] answer;
		combination = new ArrayList<>();

		for (int i = 0; i < orders.length; i++) {
			char[] orders_char = orders[i].toCharArray();
			Arrays.sort(orders_char);

			for (int index = 0; index < orders_char.length - 1; index++) {
				for (int j = 0; j < course.length; j++) {
					dfs(orders_char, index, 1, course[j], String.valueOf(orders_char[index]));
				}
			}
		}

		Map<String, Integer> map = new HashMap<>();
		for (String key : combination) {
			map.put(key, map.getOrDefault(key, 0) + 1);
		}
		List<String> keySetList = new ArrayList<>(map.keySet());
		Collections.sort(keySetList, (o1, o2) -> (map.get(o2).compareTo(map.get(o1))));

		List<String> ansList = new ArrayList<>();
		for (int i = 0; i < course.length; i++) {
			int max_value = 0;

			for (String key : keySetList) {
				if (map.get(key) >= 2 && key.length() == course[i]) {
					if (map.get(key) >= max_value) {
						ansList.add(key);
						max_value = map.get(key);
					}
				}
			}
		}
		Collections.sort(ansList);
		answer = new String[ansList.size()];
		ansList.toArray(answer);

		return answer;
	}

	public static void dfs(char[] arr, int idx, int length, int course, String str) {
		if (length == course) {
			combination.add(str);
		}

		for (int i = idx + 1; i < arr.length; i++) {
			dfs(arr, i, length + 1, course, str + arr[i]);
		}
	}
}